docs: Add the SAML intro docs (#8112)
zitadel/zitadel: docs/docs/guides/integrate/login/saml.md

				<ns2:DigestValue>WSz/EQ72RZ0DTh3DBSRCElpITqM=</ns2:DigestValue>
			</ns2:Reference>
		</ns2:SignedInfo>
		<ns2:SignatureValue>HHjsNh0OLj7...</ns2:SignatureValue>
		<ns2:KeyInfo>
			<ns2:X509Data>
				<ns2:X509Certificate>MIIDtzCCAp+gAwIBAgIUfITRQGue...</ns2:X509Certificate>
			</ns2:X509Data>
		</ns2:KeyInfo>
	</ns2:Signature>
</ns0:AuthnRequest>
```

In this SAML request:

- The **ID** (id-8LjuzBEUQFYFWjL55) is generated by the service provider, and the identity provider will respond to this ID in the SAML response.
- **IssueInstant** (2024-05-11T04:13:52Z) is the timestamp for this request, used by the identity provider for verification to ensure the request is within an acceptable time frame.
- **Destination** (https://my-instance-xtzfbc.zitadel.cloud/saml/v2/SSO) points to the identity provider's URL, ensuring the request is sent to the correct recipient.
- **AssertionConsumerServiceURL** (http://127.0.0.1:5000/acs) indicates where the response should be sent after user authentication.
- **Issuer** (https://zitadel-test.sp/metadata) is a predefined string formatted as a URL, matching what the identity provider expects.

> - **Signature** ensures the integrity and authenticity of the SAML request. It includes:

Reviewer: I would add a space or something to have a split as SignedInfo, Reference, SignatureValue and KeyInfo are inside Signature

Author: Done

- **SignedInfo** contains details about the canonicalization and signature methods.
- **Reference** points to the signed data and includes a transform and digest method.
- **SignatureValue** provides the actual digital signature.
- **KeyInfo** containing the X.509 certificate, which holds the public key used to verify the signature.

> These XMLs are stringified, encoded, and passed in the query string to the identity provider, where the user is authenticated, and the SAML response is prepared, which will be received by the SP as an encoded string. 

Reviewer: What do you mean as query string? If you mean as URL parameters, then maybe a comment to the different Bindings would be necessary as well, as the request is send as parameter or request body related to the bindings.

Reviewer: It's also dependend on the binding, as to how the signature is sent.

Author: ```suggestion
These XMLs are stringified, encoded, and transmitted according to the SAML binding used. For instance, in HTTP Redirect Binding, the request is sent as URL parameters, while in HTTP POST Binding, it is included in the request body. The signature's transmission method also depends on the binding used, ensuring the integrity and authenticity of the SAML message across different communication channels.
```

			</AuthnContext>
		</AuthnStatement>
		<AttributeStatement>
			<Attribute Name="Email" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic">
				<AttributeValue>tony.stark@gmail.com</AttributeValue>
			</Attribute>
			<Attribute Name="SurName" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic">
				<AttributeValue>Stark</AttributeValue>
			</Attribute>
			<Attribute Name="FirstName" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic">
				<AttributeValue>Tony</AttributeValue>
			</Attribute>
			<Attribute Name="FullName" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic">
				<AttributeValue>Tony Stark</AttributeValue>
			</Attribute>
			<Attribute Name="UserName" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic">
				<AttributeValue>tony.stark@gmail.com</AttributeValue>
			</Attribute>
			<Attribute Name="UserID" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic">
				<AttributeValue>260242264868201995</AttributeValue>
			</Attribute>
		</AttributeStatement>
	</Assertion>
</Response>
```

In this SAML response:

> - The **ID** (_164ba12b-6711-40e0-8ddb-55aa810f1c92) matches the ID from the SAML request (id-8LjuzBEUQFYFWjL55), allowing the service provider to verify it.

Reviewer: Which part of the content of this 2 ID's match here?

Author: ID="_164ba12b-6711-40e0-8ddb-55aa810f1c92" InResponseTo="id-8LjuzBEUQFYFWjL55"

- **InResponseTo** references the ID of the SAML request.
- **IssueInstant** (2024-06-11T04:17:41Z) is the timestamp of the response.
- **Destination** (http://127.0.0.1:5000/acs) confirms that the response is intended for the service provider's Assertion Consumer Service URL.
- **Issuer** (https://my-instance-xtzfbc.zitadel.cloud/saml/v2/metadata) is the identity provider's unique string, used to verify the response's origin.

> - **StatusCode** indicates the status of the authentication process, with a value of Success.

Reviewer: Mabye a comment that if not Successfull there was a problem with the login?

Author: ```suggestion
- **StatusCode** indicates the status of the authentication process, with a value of "Success." If the status is not "Success," it means there was a problem with the login, such as incorrect credentials or other authentication issues.
```

> - **Signature** ensures the integrity and authenticity of the SAML response. It includes:

Reviewer: Same comment about the spacing as above.

Author: Done

- **SignedInfo** contains details about the canonicalization and signature methods.
- **Reference** points to the signed data and including transforms and a digest method.
- **SignatureValue** provides the actual digital signature.
- **KeyInfo** contains the X.509 certificate, which holds the public key used to verify the signature.
- **Assertion** includes user information, such as email, surname, first name, full name, username, user ID etc.
- **Subject** contains the user's unique identifier.
- **Conditions** specify the response's validity window.
- **AuthnStatement** includes authentication details.
- **AttributeStatement** contains additional user attributes.

## SAML identity brokering

### How SAML identity brokering works

- **Initial Authentication Request**: A user attempts to access a service (SP1) that is protected by an IdP (IdP1).
- **Redirection to IdP1**: The user is redirected to IdP1 for authentication. If IdP1 trusts another IdP (IdP2) for authentication, it will redirect the user to IdP2.
- **Authentication at IdP2**: The user authenticates with IdP2, which generates a SAML assertion containing the user's identity and attributes.

> - **Assertion Transfer**: IdP2 sends the SAML assertion back to IdP1. IdP1 processes this assertion and may add additional attributes or modify the assertion based on its own policies.

Reviewer: There should be no transfer? If ZITADEL is IdP1, the Assertion from IdP2 is processed and an independent assertion is created and used as response to SP1.

Author: ```suggestion
- **Assertion Processing**: ZITADEL (IdP1) processes the SAML assertion from IdP2. ZITADEL then creates an independent SAML assertion based on the information received and its own policies.
- **Response to SP1**: ZITADEL (IdP1) sends this newly created SAML assertion to SP1, completing the authentication process.
```
